app/models/data_warehouse_export.rb

```diff
@@ -0,0 +1,33 @@
+class DataWarehouseExport < ApplicationRecord
+  EARLIEST_RANGE_FROM = Time.new(2000, 1, 1).utc.freeze
+
+  after_initialize :set_date_range
+
+  def self.generate!
+    new.tap do |export|
+      ActiveRecord::Base.transaction(isolation: :read_committed) do
```

Comment:
I'm not sure this is the right [isolation level](https://www.postgresql.org/docs/9.5/transaction-iso.html#MVCC-ISOLEVEL-TABLE) – I think we need to go one or two down to either `:serializable` or `:repeatable_read`. Read committed still allows submissions and their entries, when committed, to change under our feet.

Reply:
Really? I'm very confused now, I interpreted the following to mean that even committed changes outside of the transaction would be excluded:

> Read Committed is the default isolation level in PostgreSQL. When a transaction uses this isolation level, a SELECT query (without a FOR UPDATE/SHARE clause) sees only data committed before the query began; it never sees either uncommitted data or changes committed during query execution by concurrent transactions.

Reply:
From [Repeatable Read](https://www.postgresql.org/docs/9.5/transaction-iso.html#XACT-REPEATABLE-READ) docs:

> The Repeatable Read isolation level only sees data committed before the transaction began; it never sees either uncommitted data or changes committed during transaction execution by concurrent transactions.

Reply:
In effect, Repeatable Read will ensure that all the `Export::Anything`s operate on precisely the data that was there when the transaction block started, whereas Read Committed only isolates to each of the four queries, which would mean if a submission is committed during `Export::Anything.new(Export::Submissions::Extract.all_relevant(export.date_range)).run` that we would export its `Invoices` and `Contracts` without its `Submission`.

Reply:
Urgh, I'm really struggling with these docs. OK, I think I get it now, I've updated the isolation level to `:repeatable_read`.

Reply:
Yes, I confused myself reading about it as well, I double-took on Read Committed 😞 

Comfort yourself with the fact that you've done some good work to make transaction levels *other* than the default `:read_committed` possible in specs and since that is the default that work would have been unnecessary (as `.transaction do` would have been fine) without this change :wink: